### PR TITLE
Celltags in core

### PR DESCRIPTION
**Pull request by markellekelly**
Adds celltags to core JupyterLab. 🎉 

## References
Addresses #6952 and jupyterlab/jupyterlab-celltags#179. 

## Code changes
Creates celltags and celltags-extension packages.

## User-facing changes
Adds cell tag area to the notebook tools panel.
Users can: 
* Add tags to individual cells 
* Remove tags from individual cells
* View tags in active cell
* View tags in notebook

image

## Backwards-incompatible changes
None

Pinging @vidartf, @telamonian, @choldgraf, @isabela-pf for review and testing.

### CODE REVIEW

**Comment by choldgraf** [comment]
I think this is a great idea! I'd love to have tags work "out of the box" for jupyter book users.

Is it a copy-over from the celltags repo, or a re-implementation?

**Comment by Zsailer** [comment]
@choldgraf this is a reimplementation. 

I'll let @markellekelly provide more details. She and @isabela-pf did some user testing on the old jupyterlab/jupyterlab-celltags and found the old UI needed some simplifying. While exploring a new UI/UX, @markellekelly also took this opportunity to simplify the extension code/logic. I think this refactoring will help with future maintenance :)

**Comment by choldgraf** [comment]
I am all for a re-implementation :-)

One thing worth thinking about (doesn't have to be in this PR) is if there's value in giving people visual cues of what tags are attached to a cell without that cell being selected. In the old notebook tag system, you'd see a little word for each tag within the cell header, which made it easy to quickly skim them and see what tags were where. Now that you have to click a cell to see what tags are attached, it makes this a big clunkier.

**Comment by vidartf** [comment]
> In the old notebook tag system, you'd see a little word for each tag within the cell header, which made it easy to quickly skim them and see what tags were where.

FYI: That is only if you've selected the tags cell headers.

> Where's a good place to iterate on issues / feedback / etc if this gets into core? The jupyterlab/jupyterlab repo?

This will be the place for working on this (monorepo, yay). You seem to indicate your feature requests should happen *after* this PR, which I wholeheartedly agree with. 👍

**Comment by vidartf** [comment]
PS: Doing a quick scan of the code, it generally looks nice an clean 👍 One task that should ideally be done before merging is to add docstrings to all the public (and protected?) methods and functions.

**Comment by telamonian** [comment]
@markellekelly 

Kudos on adding dark mode styling. However, c6621b5 does not represent the right approach to dark mode theming, for the icons or otherwise. The commit should be reverted.

In jlab core, icons need to be added such that they show up in the DOM as svg nodes, not as CSS background images. This allows us to use CSS variables to directly change the styling of an icon along with the theme. Here's how you do that:

- Move your "check" icon over to `ui-components/style/icons/toolbar/check.svg`
  - just move the light mode version of the icon. Delete the dark mode version
  - don't include the icon size in the name
  - there's already an `ui-components/style/icons/toolbar/add.svg`, so you can just delete your "add" icon
- Open `check.svg` in a text editor and add `class="jp-icon3"` to it's inner `<path .../>`. 
  - there's lots of [examples](https://raw.githubusercontent.com/jupyterlab/jupyterlab/172e664aa66d8c9be89960c575d24dc4051e7bb9/packages/ui-components/style/icons/toolbar/add.svg) of svgs appropriately annotated with `class` in `ui-components`
- In your actual code (eg in the [`buildTag`](https://github.com/jupyterlab/jupyterlab/pull/7407/files#diff-1de68d162a8b92405f7a0d77be1d0f84R20) functions), import `defaultIconRegistry` from `ui-components` and create your icons using the `defaultIconRegistry.icon` method
  - here's an example use:
https://github.com/jupyterlab/jupyterlab/blob/172e664aa66d8c9be89960c575d24dc4051e7bb9/packages/apputils-extension/src/index.ts#L168-L173
  - given that your icon file is named `check.svg`, the name argument for `.icon` should just be `'check'`
  - `center` and `kind` are optional arguments that control the resulting icon's style. For your case, you should probably not use these. Instead, `.icon` is able to take any standard CSS property as an argument, so you can style your icon that way
    - alternatively, you can look at [`ui-components/src/style/icon.ts`](https://github.com/jupyterlab/jupyterlab/blob/172e664aa66d8c9be89960c575d24dc4051e7bb9/packages/ui-components/src/style/icon.ts) for the valid values of `kind`, and possibly set up your own using `typestyle`

**Comment by telamonian** [comment]
For general non-icon light mode/dark mode styling, importing/requiring the theme manager is also the wrong approach. Instead, you can just prefix the selector of any light mode specific CSS with

```css
[data-jp-theme-light='true']
```

and prefix the selector of any dark mode specific CSS with

```css
[data-jp-theme-light='false']
```

Here's an existing example of this in jlab core:

https://github.com/jupyterlab/jupyterlab/blob/172e664aa66d8c9be89960c575d24dc4051e7bb9/packages/rendermime/style/base.css#L420-L426

**Comment by markellekelly** [comment]
@telamonian thanks!! should be fixed now

File: packages/celltags-extension/package.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  "scripts": {
+    "build": "tsc",
+    "clean": "rimraf lib && rimraf tsconfig.tsbuildinfo",
+    "prepare": "npm run clean && npm run build",
+    "prepublishOnly": "npm run build",
+    "watch": "tsc -b --watch"
+  },
+  "dependencies": {
+    "@jupyterlab/application": "^2.0.0-alpha.1",
+    "@jupyterlab/apputils": "^2.0.0-alpha.1",
```
Comment (telamonian) [inline packages/celltags-extension/package.json]: Is this dependency still in use anywhere in your code? If not, you should remove it from the `package.json`

File: packages/celltags/package.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  "scripts": {
+    "build": "tsc -b",
+    "clean": "rimraf lib",
+    "prepare": "npm run clean && npm run build",
+    "prepublishOnly": "npm run build",
+    "watch": "tsc -b --watch"
+  },
+  "dependencies": {
+    "@jupyterlab/application": "^2.0.0-alpha.1",
+    "@jupyterlab/apputils": "^2.0.0-alpha.1",
```
Comment (telamonian) [inline packages/celltags/package.json]: here too

**Comment by markellekelly** [comment]
Note: due to onActiveCellMetadataChanged not detecting changes within the 'tags' list (only changes in whether it exists), the separate celltags extension and in-panel cell metadata editor do not always update to reflect changes in tags immediately.

File: packages/celltags/package.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    "prepare": "npm run clean && npm run build",
+    "prepublishOnly": "npm run build",
+    "watch": "tsc -b --watch"
+  },
+  "dependencies": {
+    "@jupyterlab/application": "^2.0.0-alpha.1",
+    "@jupyterlab/cells": "^2.0.0-alpha.1",
+    "@jupyterlab/notebook": "^2.0.0-alpha.1",
+    "@jupyterlab/ui-components": "^2.0.0-alpha.1",
+    "@phosphor/widgets": "^1.9.0",
```
Comment (saulshanabrook) [inline packages/celltags/package.json]:
```suggestion
    "@lumino/widgets": "^1.9.0",
```

**Comment by Zsailer** [comment]
Tested out this extension today. 

I'm seeing an issue when creating a tag. When I click on the starter tag, the text area for naming the tag doesn't immediately come into focus. I have to double click to start typing. A single click puts me in a weird state where typing doesn't do anything.

**Comment by Zsailer** [comment]
@markellekelly, can you check the box on the right side of the PR to allow maintainers to edit your branch directly?

**Comment by Zsailer** [comment]
Thanks for making those changes, @markellekelly. 

A couple new issues 😆 

1. After naming my tag, the "+" button doesn't add a tag. I have to hit enter.
2. Tags with spaces in the name create multiple tags—one for each word.

**Comment by markellekelly** [comment]
@Zsailer I changed the "+" button to serve as an alternative to hitting Enter. Separating tags by spaces and commas is currently the intended behavior, but I can change this if people find it confusing or inconvenient.

**Comment by Zsailer** [comment]
Great, thanks @markellekelly. I'll work with @telamonian to review this in the coming days. Thanks for all your hard work!

**Comment by choldgraf** [comment]
Yessss this is very exciting!

**Comment by telamonian** [comment]
@markellekelly Nice work!

**Pull request merged**